File: L3/S6/RESEAUX/TP4-DNS/ressources/DNSsimple.c
/*
 * Original de Laurent NOE ~ 2009
 * Mise à jour du traitement des erreurs, des includes, et de l'affichage - Arthur ROLAND & Rudy SEYS - Mars 2017
 * Mise à jour des noms/adresses des serveurs DNS de la fac, affichage et commentaires - Laurent NOE - Janvier 2019
 * Ajout de quelques commentaires pour autres serveurs fac et exterieurs               - Laurent NOE - Janvier 2020
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <string.h>
#include <netdb.h>
#include <unistd.h>
#include <assert.h>

#define PROGRAM_CALL_DNS

#define QUERYLENGTH 29
#define NS_PACKETSZ 512

/* NOM/ADRESSE-IP + PORT  du serveur DNS qui recevra notre requete */
#define HOSTDESTNAME "reserv1.univ-lille1.fr" /*  "reserv2.univ-lille1.fr" */
        /* utiliser "1.1.1.1" / "9.9.9.9", ou celui déclaré dans le fichier /etc/resolv.conf de chez vous, si les suivants ne sont pas joignables.
         * utiliser "reserv1.univ-lille1.fr" (sinon "reserv2.univ-lille1.fr") à l'université pour avoir les autorites/additionels sur "www.lifl.fr".
         * utiliser "dns1.unv-lille.fr" (sinon "dns2.univ-lille.fr") sur l'université (mais sans autorite/additionel)
	 */
#define HOSTDESTADDR "193.49.225.15" /* "193.49.225.90" , "1.1.1.1" */
#define PORT     53

int main(void)
{
    int sock = 0;
    struct sockaddr_in name;
    char query[QUERYLENGTH] = {
        (char) 0x08, (char) 0xbb, (char) 0x01, (char) 0x00, /* a) 12 octets d'entete : identifiant de requete/parametres [RFC1035, 4.1.1] */
        (char) 0x00, (char) 0x01, (char) 0x00, (char) 0x00,
        (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00,
        (char) 0x03, (char) 0x77, (char) 0x77, (char) 0x77, /* b) QUESTION                     [RFC1035, partie 4.1.2] */
        (char) 0x04, (char) 0x6c, (char) 0x69, (char) 0x66, /* b.1) QNAME : "3www4lifl2fr0"    [RFC1035, partie 3.1] */
        (char) 0x6c, (char) 0x02, (char) 0x66, (char) 0x72,
        (char) 0x00,
        (char) 0x00, (char) 0x01,                           /* b.2) QTYPE : A (a host address) [RFC1035, partie 3.2.3] */
        (char) 0x00, (char) 0x01                            /* b.3) QCLASS : IN (the Internet) [RFC1035, partie 3.2.4] */
    };

    char buffer[NS_PACKETSZ];

    /* recvfrom */
    ssize_t len;
    int i;
    struct sockaddr_in addrClient;
    socklen_t addrClientlen = 0;

#ifdef PROGRAM_CALL_DNS
    struct hostent *hp = NULL;
    char **p = NULL;
#endif

    memset(&addrClient, 0, sizeof(struct sockaddr_in));
    memset(&name, 0, sizeof(struct sockaddr_in));


    /* 1) obtenir l'adresse IP du serveur DNS ... en utilisant une requete DNS ??! */
#ifdef PROGRAM_CALL_DNS
    fprintf(stdout, " recherche de l'IP, pour le nom de l'hote de destination \"%s\" ... ",HOSTDESTNAME);
    hp = gethostbyname(HOSTDESTNAME);   /* convertir NOM SYMBOLIQUE en @IP */
    if (hp == NULL) {
        perror("[erreur] - gethostbyname ");
        return EXIT_FAILURE;
    }
    fprintf(stdout, "[ok]\n");

    /* affichage */
    fprintf(stdout, " - h_name : \"%s\"\n", hp->h_name);

    for (p = hp->h_aliases; *p != NULL; p++)
        fprintf(stdout, " - h_aliases : \"%s\"\n", *p);

    fprintf(stdout, " - h_addrtype : %s\n",
            (hp->h_addrtype ==
             AF_INET) ? "AF_INET" : ((hp->h_addrtype ==
                                      AF_INET6) ? "AF_INET6" : "unknown"));
    fprintf(stdout, " - h_length : %d\n", hp->h_length);

    for (p = hp->h_addr_list; *p != NULL; p++)
        fprintf(stdout, " - h_addr_list : \"%x.%x.%x.%x\"\n",
                *(*p) & 0xff, *(*p + 1) & 0xff, *(*p + 2) & 0xff,
                *(*p + 3) & 0xff);

    fprintf(stdout, " - h_addr : \"%x.%x.%x.%x\"\n",
            *(*hp->h_addr_list) & 0xff, *(*hp->h_addr_list + 1) & 0xff,
            *(*hp->h_addr_list + 2) & 0xff,
            *(*hp->h_addr_list + 3) & 0xff);

    /* 2) creation d'un socket udp */
    fprintf(stdout, " creation du socket en mode DGRAM (UDP) ... ");
    sock = socket(PF_INET, SOCK_DGRAM, 0);
    if (sock < 0) {
        perror("[erreur] - socket ");
        return EXIT_FAILURE;
    }

    fprintf(stdout, "[ok]\n");

    /* 3) preparation du socket d'envoi */
    name.sin_family = (short unsigned int) hp->h_addrtype;
    name.sin_port = htons(PORT);
    memcpy(&name.sin_addr.s_addr, hp->h_addr_list[0], (long unsigned int) hp->h_length);        /* <- memcpy + inversion!!! */

#else /* PROGRAM_CALL_DNS */

    /* 3) autre possibilite : on connait l'@IP du serveur DNS et c'est la solution la plus logique (le probleme de "poule et d'oeuf" disparait)  */
    fprintf(stdout, " conversion de l'IP de l'hote de destination \"%s\" en binaire ... ",HOSTDESTADDR);
    name.sin_family = AF_INET;
    name.sin_port = htons(PORT);
    if (inet_pton(AF_INET, HOSTDESTADDR, &name.sin_addr.s_addr) < 0) {
        perror("[erreur] - inet_pton ");
    }
    fprintf(stdout, "[ok]\n");

#endif /* PROGRAM_CALL_DNS */

    /* 4) envoi du message */
    fprintf(stdout, " envoi du message ... ");
    if ((len = sendto
        (sock, query, QUERYLENGTH, 0, (struct sockaddr *) &name,
         sizeof(struct sockaddr_in))) < 0) {
        perror("[erreur] - sendto ");
        return EXIT_FAILURE;
    }
    fprintf(stdout, "[ok]\n longueur du message envoye : %lu\n", len);

    /*
     * 5) reception du message
     */
    fprintf(stdout, " reception du message ... ");
    if ((len =
         recvfrom(sock, buffer, NS_PACKETSZ, 0,
                  (struct sockaddr *) &addrClient, &addrClientlen)) < 0) {
        perror("[erreur] - recvfrom ");
        return EXIT_FAILURE;
    }
    fprintf(stdout, "[ok]\n longueur du message recu : %lu\n", len);

    /* fermeture de la liaison. */
    close(sock);

    /* affichage de la reponse */

    /* affichage %c */
    for (i = 0; i < len; i++) {
      fprintf(stdout, " %c ", buffer[i] > 31 ? buffer[i] : '.');
        if (!((i + 1) % 16))
            fprintf(stdout, "\n");
    }
    fprintf(stdout, "\n");

    /* affichage %h */
    for (i = 0; i < len; i++) {
        fprintf(stdout, "%.2x ", (int) (buffer[i] & 0xff));
        if (!((i + 1) % 16))
            fprintf(stdout, "\n");
    }
    fprintf(stdout, "\n");

    return EXIT_SUCCESS;
}

File: L3/S6/RESEAUX/TP4-DNS/ressources/DNSsimple.java
/*
 * Original de Laurent NOE ~ 2009
 * Mise à jour des noms/adresses des serveurs DNS de la fac, affichage et commentaires - Laurent NOE - Janvier 2019
 * Ajout de quelques commentaires pour autres serveurs fac et exterieurs               - Laurent NOE - Janvier 2020
 */

import java.io.*;
import java.net.*;
import java.util.*;

/*
 * envoi question DNS / reception reponse et affichage hexa rapide ...
 */

public class DNSsimple
{
    public static void main(String[] args)
    {
        /* NOM/ADRESSE-IP + PORT  du Serveur DNS qui recevra notre requete */
        String HOSTDESTNAME = "reserv1.univ-lille1.fr";
        /* utiliser "1.1.1.1" / "9.9.9.9", ou celui déclaré dans le fichier /etc/resolv.conf de chez vous, si les suivants ne sont pas joignables.
         * utiliser "reserv1.univ-lille1.fr" (sinon "reserv2.univ-lille1.fr") à l'université pour avoir les autorites/additionels sur "www.lifl.fr".
         * utiliser "dns1.unv-lille.fr" (sinon "dns2.univ-lille.fr") sur l'université (mais sans autorite/additionel)
         */
        int PORT = 53;



        
        byte[] message = {(byte) 0x08, (byte) 0xbb,  (byte) 0x01, (byte) 0x00, /* a) 12 octets d'entete : identifiant de requete/parametres [RFC1035, 4.1.1] */
                          (byte) 0x00, (byte) 0x01,  (byte) 0x00, (byte) 0x00,
                          (byte) 0x00, (byte) 0x00,  (byte) 0x00, (byte) 0x00,
                          (byte) 0x03, (byte) 0x77,  (byte) 0x77, (byte) 0x77, /* b) QUESTION                     [RFC1035, partie 4.1.2] */
                          (byte) 0x04, (byte) 0x6c,  (byte) 0x69, (byte) 0x66, /* b.1) QNAME : "3www4lifl2fr0"    [RFC1035, partie 3.1] */
                          (byte) 0x6c, (byte) 0x02,  (byte) 0x66, (byte) 0x72,
                          (byte) 0x00,
                          (byte) 0x00, (byte) 0x01,                            /* b.2) QTYPE : A (a host address) [RFC1035, partie 3.2.3] */
                          (byte) 0x00, (byte) 0x01 };                          /* b.3) QCLASS : IN (the Internet) [RFC1035, partie 3.2.4] */


        /* 1) obtenir l'adresse IP du serveur DNS ... en utilisant une requete DNS ??! */
        System.err.print(" recherche de l'IP, pour le nom de l'hote de destination \""+HOSTDESTNAME+"\" ... ");
        InetAddress destination;
        try {
            destination = InetAddress.getByName(HOSTDESTNAME);
        } catch (Exception e) {
            System.err.println("[error] :" +  e.getMessage());
            return;
        }
        System.err.println("[ok]");

        /* 2) creation d'un DatagramPacket pour l'envoi de la question DNS */
        System.err.println(" creation du Datagrampacket, taille du message : "+message.length  );
        DatagramPacket dp = new DatagramPacket(message,message.length,destination,PORT);

        /* 3) creation d'un DatragramSocket (port au choix ) */
        System.err.print(" creation du Datagramsocket  ... ");
        DatagramSocket ds ;
        try {
            ds = new DatagramSocket() ;
        } catch (Exception e) {
            System.err.println("[error] :" +  e.getMessage());
            return;
        }
        System.err.println("[ok]");

        /* 4) et envoi du packet */
        System.err.print(" envoi du message ... ");
        try {
            ds.send(dp);
        } catch (Exception e) {
            System.err.println("[error] :" +  e.getMessage());
            return;
        }
        System.err.println("[ok]");

        /* 5) reception du packet */
        dp = new DatagramPacket(new byte[512],512);
        System.err.print(" reception du message ... ");
        try {
            ds.receive(dp);
        } catch (Exception e) {
            System.err.println("[error] :" +  e.getMessage());
            return;
        }
        System.err.println("[ok]");

        /* affichage complet du packet recu (pas tres lisible ...) */
        byte[] rec = dp.getData();
        System.out.println(" longueur du message recu : " + dp.getLength());

        /* affichage des bytes (et ascii associes) */
        for(int i = 0; i < dp.getLength(); i++) {
            /* NOTE :
             *   utiliser au besoin l'expression "rec[i] & 0xff" pour obtenir une valeur
             *   entre 0 et 255 (et pas entre -128 et +127).
             */

            System.out.print(" "+String.format("%02X",rec[i]));
    
            if (((i+1)%16 == 0) || (i+1 == dp.getLength())) {

                /* ceci pour afficher les caracteres ascii aprec l'hexa */
                /* >>> */
                System.out.print("\t");

                for (int j = 0 ; j < ((i+1) & 15); j++) {
                    System.out.print("     ");
                }
                for (int j = i & ~15; j <= i; j++) {
                    if (rec[j] > 31) {
                        System.out.print((char)rec[j]);
                    } else {
                        System.out.print(".");
                    }
                }
                /* <<< */

                System.out.println("");
            }
        }
    }
}



